libgadget/sfr_eff.c

```diff
@@ -25,40 +25,61 @@
 #include "sfr_eff.h"
 #include "cooling.h"
 #include "slotsmanager.h"
-#include "timestep.h"
-#include "treewalk.h"
 #include "winds.h"
-
+/*Only for the star slot reservation*/
+#include "forcetree.h"
+#include "timestep.h"
+#include "domain.h"
 /*Cooling only: no star formation*/
 static void cooling_direct(int i);
 
 static void cooling_relaxed(int i, double egyeff, double dtime, double trelax);
 
 static int sfreff_on_eeqos(int i);
-static int make_particle_star(int i, double mass_of_star);
-static int starformation(int i, double *localsfr, double * sum_sm, double *sum_mass_stars);
+static int make_particle_star(int child, int parent, int placement);
+static int starformation(int i, double *localsfr, double * sum_sm);
 static int quicklyastarformation(int i);
 static double get_sfr_factor_due_to_selfgravity(int i);
 static double get_sfr_factor_due_to_h2(int i);
 static double get_starformation_rate_full(int i, double dtime, MyFloat * ne_new, double * trelax, double * egyeff);
 static double find_star_mass(int i);
+/*Get enough memory for new star slots. This may be excessively slow! Don't do it too often.*/
+static int * sfr_reserve_slots(int * NewStars, int NumNewStar);
+
 
 /* cooling and star formation routine.*/
 void cooling_and_starformation(void)
 {
     if(!All.CoolingOn)
         return;
 
-    int i;
+    /*This is a queue for the new stars and their parents, so we can reallocate the slots after the main cooling loop.*/
+    int * NewStars = NULL;
+    int * NewParents = NULL;
+    int NumNewStar = 0;
+    size_t *nqthrsfr = ta_malloc("nqthrsfr", size_t, All.NumThreads);
+    int **thrqueuesfr = ta_malloc("thrqueuesfr", int *, All.NumThreads);
+    int **thrqueueparent = ta_malloc("thrqueueparent", int *, All.NumThreads);
+
+    /*Need to capture this so that when NumActiveParticle increases during the loop
+     * we don't add extra loop iterations on particles with invalid slots.*/
+    const int nactive = NumActiveParticle;
+
+    if(All.StarformationOn) {
+        NewStars = mymalloc("NewStars", nactive * sizeof(int) * All.NumThreads);
+        gadget_setup_thread_arrays(NewStars, thrqueuesfr, nqthrsfr, nactive, All.NumThreads);
+        NewParents = mymalloc2("NewParents", nactive * sizeof(int) * All.NumThreads);
+        gadget_setup_thread_arrays(NewParents, thrqueueparent, nqthrsfr, nactive, All.NumThreads);
```

Comment:
Could be easier if 
1. gadget_setup_thread_arrays allocates thrqueueparent with malloc and returns the per thread array directly. Then the frees later can be regular free(no ordering requirements), and you can hide a few ta_mallocs.

2. gadget_setup_thread_arrays also has almost enough information to do the mymalloc stuff, but it appears you need to use two different allocators because they have to be freed in different orders?

Reply:
I see the memory usage can be NumPart * NThreads, which can be huge. There are two ways of dealing with this:
- If we use a static scheduling policy, then the total size of NewStars can be nactive, rather than each has nactive. But the thread loads can be very imbalanced. We can try to schedule by interval rather than by chunk (probably need to manually write a for loop inside omp parallel.

- Use a single queue. What if you use an atomic capture to advance a single queue pointer? Would it be a lot slower than having one queue per thread?

Reply:
This is true, but never the peak memory because we have a similar allocation whenever we use gadget_setup_thread_arrays(). So every time the treewalk gets initialised we get something like this. It is also only one int * NumPart * NumThreads which is not too much - I can put a report_memory_usage() around here so we know if this changes.

Reply:
The single queue with the atomic capture would be very much slower, unfortunately.

Reply:
Also static scheduling is slower by a factor of several: I checked this.

Reply:
By static scheduling you mean using openmp's static keyword? It uses same number of equal sized chunks as number of threads, and should have had bad imbalance between threads.

I was thinking more like this

for(i = tid; i < NumPart; i += nthreads) {

}

This would mean each thread will get into the hot regions equal likely, and we can also bound the number of particles added to the queue. I guess if this does not slow down badly here we can also shrink other places where such queue is needed.

The memory blow up is going to be a problem, if we are working with 64 threads this temporary queue is going to be bigger than the particle data.

Reply:
Ah, that I didn't think of doing that! It's a good idea and could be done elsewhere as well. I'll check the speed change.

Reply:
We might break the pre-fetcher doing this though: even the dynamic scheduler is mostly contiguous.

Reply:
is the new scheme slower? Yes the caches are likely per core and prefetching is going to be broken. But I am betting we do so much calculation in start forming particles, missing some caches is not a big problem.

Reply:
It is a little slower but not noticeably so. I think you would win your bet.